### PR TITLE
Refactored code base - version v1

### LINKED ISSUE #53: Improve/Refactor Dependency Watchdog (v2)
# Stories
- As developer I want a robust, scalable, and maintainable code base, so that issues are minimised and maintenance effort improves.

# Motivation
We made the following observations:
- Bringing down KCM when it's not required #35
- Inability to handle kubeconfig rotation for all edge scenarios #36
- Lack of Test automation to be able to release with confidence (current code coverage of prober is 9%)
- Just keeping the code base updated with evolution of Kubernetes for e.g. adapting to stories around https://github.com/gardener/gardener/issues/4251
- As a developer the current code base also is pretty convoluted to maintain and requires sufficient efforts for someone new to make headway into the code.

All these observations, when put together, make it apparent that the fixes provided in the recent past are more band-aids that will back-fire with more and more Gardener adoption. DWD, being a critical component for the health of the Gardener-managed control planes, should be redesigned/refactored/refined/adapted to improve robustness, scalability, and maintainance effort.

# Acceptance Criteria
- [ ] Documentation/Proposal
  - [x] Create a design document for prober
  - [x] Enhance the design to include weeder which was previously called endpoint command
  - [x] Re-structure README and create multiple markdown files which the consumers can reference
  - [ ] Add example configurations for weeder and prober

- [x] Prober
  - [x] Implement new Prober design
  - [x] Write Unit tests - target 90%+ functional coverage
  - [x] Programmatically bring up a vanilla KIND cluster for unit tests
  - [x] Look at all Open/Close issues on DWD w.r.t prober and verify that all issues are fixed in the new code
  - [x] Work on gardener PR for https://github.com/gardener/gardener/issues/5878

- [ ] Weeder
  - [x] Finalize the new design
  - [x] Implement new weeder design
  - [x] Unit and integration tests - target 90% functional coverage
  - [x] Look at all Open/Close issues on DWD w.r.t weeder and verify that all issues are fixed in the new code
  - [ ] Work on gardener PR for https://github.com/gardener/gardener/issues/5878

- [ ] CI/CD
  - [x] Configure Prow jobs for dependency watchdog
  - [x] Enhance the Makefile and add necessary scripts under `/hack` (take inspiration from g/g)
  - [x] Identify tools that are required for make targets and create a `tools.mk` for it.

- [ ] Metrics
  - [ ] Identify metrics that should be exposed out of prober/weeder
  - [ ] Identify the dashboard on Graphana where these metrics will be pushed and create relevant graphs/charts.

- [ ] Botanist: Configuration change Issue/PR
  - [ ] Create an issue on gardener/gardener and raise a PR to change the prober configuration in operations/botanist

# Enhancement/Implementation Proposal (optional)
- Relook at how we can add more components to the dependency graph of resources as well as how we can handle secret rotation (https://github.com/gardener/dependency-watchdog/issues/38).
- We should capture the change in a proposal doc which can be shared with interested parties and community at large.
- We should do a community demo to showcase and share the change and its benefits with the community.

### PR DESCRIPTION
**Pull request by unmarshall**
**What this PR does / why we need it**:
Current version of dependency watchdog has a lot of issues and shortcomings. There was a dire need to refactor the codebase. This PR addresses many of these concerns.

**Which issue(s) this PR fixes**:
Fixes #53 

**Special notes for your reviewer**:

**Release note**:

```improvement operator
This PR introduces backward incompatible changes to the API/Config. Consumers should adapt to the new changes.
```

### CODE REVIEW

**Review by rishabh-11 — APPROVED** [review]

**Review by himanshu-kun — APPROVED** [review]
/lgtm

File: api/prober/types.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	// DependentResourceInfos dependent resources that should be considered for scaling in case the shoot control API server cannot be reached via external domain
+	DependentResourceInfos []DependentResourceInfo `json:"dependentResourceInfos"`
+}
+
+// DependentResourceInfo captures a dependent resource which should be scaled
+type DependentResourceInfo struct {
+	// Ref identifies a resource
+	Ref *autoscalingv1.CrossVersionObjectReference `json:"ref"`
+	// ShouldExist should be true if this resource should be present. If the resource is optional then it should be false.
+	ShouldExist *bool `json:"shouldExist"`
```
Comment (shreyas-s-rao) [inline api/prober/types.go:51]: Can we change this field name to `Optional`, which is in line with many other Kubernetes resource definitions? In which case, the code needs to be changed to reflect this flipping of boolean role here.

File: example/01-dwd-prober-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
+apiVersion: apps/v1
```
Comment (shreyas-s-rao) [inline example/01-dwd-prober-deployment.yaml:1]: Please rename the files in `example` directory to:
```
01-dwd-prober-configmap.yaml
02-dwd-weeder-configmap.yaml
03-dwd-prober-deployment.yaml
04-dwd-weeder-deployment.yaml
```
This maintains the correct order of components to be deployed.

**Review by shreyas-s-rao — COMMENTED** [review]
@unmarshall @rishabh-11 @himanshu-kun Thanks a lot for working on refactoring the DWD code!
After an out-of-band code walkthrough from @unmarshall , I don't see any changes required to the functionality of DWD, but just a few minor changes to the configuration API before it's integrated into gardener. PTAL

File: api/weeder/types.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// Config provides typed access weeder configuration
+type Config struct {
+	// WatchDuration Duration for which all dependent pods for a service under surveillance will be watched after the service has recovered.
+	// If the dependent pods have not transitioned to CrashLoopBackOff in this duration then it is assumed that they will not enter that state.
+	WatchDuration *metav1.Duration `json:"watchDuration,omitempty"`
+	// ServicesAndDependantSelectors is a map whose key is the service name and the value is a DependantSelectors
+	ServicesAndDependantSelectors map[string]DependantSelectors `json:"servicesAndDependantSelectors"`
+}
+
+// DependantSelectors encapsulates LabelSelector's used to identify dependants for a service
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply api/weeder/types.go:30]: Added one liner for it.

File: api/prober/types.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	ProbeInterval *metav1.Duration `json:"probeInterval,omitempty"`
+	// InitialDelay is the initial delay in running a probe for the first time
+	InitialDelay *metav1.Duration `json:"initialDelay,omitempty"`
+	// ProbeTimeout is the timeout that is set on the client which is used to reach the shoot control plane API server
+	ProbeTimeout *metav1.Duration `json:"probeTimeout,omitempty"`
+	// SuccessThreshold is the number of consecutive times a probe is successful to ascertain that the probe is healthy
+	SuccessThreshold *int `json:"successThreshold,omitempty"`
+	// FailureThreshold is the number of consecutive times a probe is unsuccessful to ascertain that the probe is unhealthy
+	FailureThreshold *int `json:"failureThreshold,omitempty"`
+	// InternalProbeFailureBackoffDuration backoff duration if the internal probe is unhealthy, before attempting again
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply api/prober/types.go:38]: Corrected.

File: api/prober/types.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	InternalProbeFailureBackoffDuration *metav1.Duration `json:"internalProbeFailureBackoffDuration,omitempty"`
+	// BackoffJitterFactor jitter with which a probe is run
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply api/prober/types.go:40]: Corrected.

File: api/prober/types.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	BackoffJitterFactor *float64 `json:"backoffJitterFactor,omitempty"`
+	// DependentResourceInfos dependent resources that should be considered for scaling in case the shoot control API server cannot be reached via external domain
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply api/prober/types.go:42]: Corrected.

File: api/prober/types.go (hunk 1 above)
Comment (unmarshall) [reply api/prober/types.go:51]: Done.

File: example/01-dwd-prober-deployment.yaml (hunk 2 above)
Comment (unmarshall) [reply example/01-dwd-prober-deployment.yaml:1]: Changed.

File: api/weeder/types.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+// limitations under the License.
+
+package weeder
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Config provides typed access weeder configuration
+type Config struct {
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply api/weeder/types.go:22]: We have a hard dependency on `Cluster` custom resource and we derive the shoot namespace from it and create a prober for every shoot cluster for which an event is sent to the cluster controller. So we have a hard dependency to gardener in the code which today makes it difficult to use the Prober outside of gardener. Unless we change that a namespace selector would not add much value. WDYT?

File: api/weeder/types.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+	// WatchDuration Duration for which all dependent pods for a service under surveillance will be watched after the service has recovered.
+	// If the dependent pods have not transitioned to CrashLoopBackOff in this duration then it is assumed that they will not enter that state.
+	WatchDuration *metav1.Duration `json:"watchDuration,omitempty"`
+	// ServicesAndDependantSelectors is a map whose key is the service name and the value is a DependantSelectors
+	ServicesAndDependantSelectors map[string]DependantSelectors `json:"servicesAndDependantSelectors"`
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply api/weeder/types.go:27]: Yes that is one way to look at it, however doing this makes it slightly inconvenient to use it internally. Right now it unmarshals to a map with the key as the name of the service/endpoint and the value as dependant pod selectors. Ideally an API struct should be kept separate from the internal struct and that provides freedom to change the API struct to be more usage friendly and clearer and the internal struct can then be optimised for implementation. We currently do not have such a distinction. However, we can take it up at a later time. I hope that is fine with you.

**Pull request merged**